clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -0,0 +1,248 @@
+import org.apache.kafka.common.KafkaException;
+
+import org.apache.kafka.common.security.kerberos.KerberosNameParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SaslClientAuthenticator implements Authenticator {
+
+    public enum SaslState {
+        INITIAL, INTERMEDIATE, COMPLETE, FAILED
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(SaslClientAuthenticator.class);
+
+    private final Subject subject;
+    private final String servicePrincipal;
+    private final String host;
+    private final String node;
+    private final KerberosNameParser kerberosNameParser;
+
+    // assigned in `configure`
+    private SaslClient saslClient;
+    private String clientPrincipalName;
+    private TransportLayer transportLayer;
+
+    // buffers used in `authenticate`
+    private NetworkReceive netInBuffer;
+    private NetworkSend netOutBuffer;
+
+    private SaslState saslState = SaslState.INITIAL;
+
+    public SaslClientAuthenticator(String node, Subject subject, String servicePrincipal, String host, KerberosNameParser kerberosNameParser) throws IOException {
+        this.node = node;
+        this.subject = subject;
+        this.host = host;
+        this.servicePrincipal = servicePrincipal;
+        this.kerberosNameParser = kerberosNameParser;
+    }
+
+    public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder, Map<String, ?> configs) throws KafkaException {
+        try {
+            this.transportLayer = transportLayer;
+
+            // determine client principal from subject.
+            Principal clientPrincipal = subject.getPrincipals().iterator().next();
+            this.clientPrincipalName = kerberosNameParser.parse(clientPrincipal.getName()).toString();
+            this.saslClient = createSaslClient();
+        } catch (Exception e) {
+            throw new KafkaException("Failed to configure SaslClientAuthenticator", e);
+        }
+    }
+
+    private SaslClient createSaslClient() {
+        try {
+            return Subject.doAs(subject, new PrivilegedExceptionAction<SaslClient>() {
+                public SaslClient run() throws SaslException {
+                    String[] mechs = {"GSSAPI"};
+                    LOG.debug("Creating SaslClient: client={};service={};serviceHostname={};mechs={}",
+                        clientPrincipalName, servicePrincipal, host, Arrays.toString(mechs));
+                    return Sasl.createSaslClient(mechs, clientPrincipalName, servicePrincipal, host, null,
+                            new ClientCallbackHandler());
+                }
+            });
+        } catch (Exception e) {
+            throw new KafkaException("Failed to create SASL client", e);
+        }
+    }
+
+
+    public void authenticate() throws IOException {
+        if (netOutBuffer != null && !flushNetOutBuffer()) {
+            transportLayer.addInterestOps(SelectionKey.OP_WRITE);
+            return;
+        }
```

Comment:
That's right. If we are still waiting for a new token to be completely received, we will need to turn off OP_WRITE.